modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/stat/IgniteStatisticsConfigurationManager.java

```diff
@@ -0,0 +1,518 @@
+                }
+
+                // Drop statistics
+                Set<String> newCols = Arrays.stream(newCfg.columns())
+                    .map(StatisticsColumnConfiguration::name)
+                    .collect(Collectors.toSet());
+
+                Set<String> colsToRemove = oldCfg != null ?
+                    Arrays.stream(oldCfg.columns())
+                        .map(StatisticsColumnConfiguration::name)
+                        .filter(c -> !newCols.contains(c))
+                        .collect(Collectors.toSet()) :
+                    Collections.emptySet();
+
+                if (log.isDebugEnabled()) {
+                    log.debug("Remove local statistics [key=" + newCfg.key() +
+                        ", columns=" + colsToRemove + ']');
+                }
+
+                LocalStatisticsGatheringContext gCtx = gatherer.gatheringInProgress(newCfg.key());
+
+                if (gCtx != null) {
+                    gCtx.future().thenAccept((v) -> {
+                        repo.clearLocalStatistics(newCfg.key(), colsToRemove);
+                        repo.clearLocalPartitionsStatistics(newCfg.key(), colsToRemove);
+                    });
+                }
+                else {
+                    repo.clearLocalStatistics(newCfg.key(), colsToRemove);
+                    repo.clearLocalPartitionsStatistics(newCfg.key(), colsToRemove);
+                }
+            }
+            else {
+                // Update statistics
+                GridH2Table tbl = schemaMgr.dataTable(newCfg.key().schema(), newCfg.key().obj());
+
+                GridCacheContext cctx = tbl.cacheContext();
+
+                Set<Integer> parts = cctx.affinity().primaryPartitions(
+                    cctx.localNodeId(), cctx.affinity().affinityTopologyVersion());
+
+                Column[] cols = IgniteStatisticsHelper.filterColumns(tbl.getColumns(), newCfg.columns());
+
+                LocalStatisticsGatheringContext ctx = gatherer
+                    .collectLocalObjectsStatisticsAsync(tbl, cols, parts, newCfg.version());
+
+                ctx.future().thenAccept((v) -> {
+                    onFinishLocalGathering(newCfg.key(), parts);
+                });
+            }
+        }
+    }
+
+    /** */
+    private void onDropColumns(GridH2Table tbl, List<String> cols) {
+        assert !F.isEmpty(cols);
+
+        dropStatistics(Collections.singletonList(
+            new StatisticsTarget(
+                tbl.getSchema().getName(),
+                tbl.getName(),
+                cols.toArray(EMPTY_STRING_ARR)
+            )
+        ));
+    }
+
+    /** */
+    private void onDropTable(String schema, String name) {
+        assert !F.isEmpty(schema) && !F.isEmpty(name) : schema + ":" + name;
+
+        dropStatistics(Collections.singletonList(new StatisticsTarget(schema, name)));
+    }
+
+    /** */
+    private void onFinishLocalGathering(StatisticsKey key, Set<Integer> partsToAggregate) {
```

Comment:
All the time we call it with cfg.key() and read configuration by key from distrMS. If it was done to get actual version - lets get actual set of partition too, if not - why not to pass the actual configuration to it?

Reply:
Will be changed with drop stat refactoring